include/deal.II/matrix_free/vector_data_exchange.h

```diff
@@ -247,6 +247,269 @@ namespace internal
+
+        /**
+         * The set of indices that appear for an IndexSet that is a subset of a
+         * larger set for each rank in a compressed manner.
+         */
+        std::pair<std::vector<unsigned int>,
+                  std::vector<std::pair<unsigned int, unsigned int>>>
+          ghost_indices_subset_data;
+
+        /**
+         * An array that contains information which processors my ghost indices
+         * belong to, at which offset and how many those indices are
+         */
+        std::vector<
+          std::pair<unsigned int, std::pair<unsigned int, unsigned int>>>
+          ghost_targets_data;
+
+        /**
+         * The set of processors and length of data field which send us their
+         * ghost data.
+         *
+         * @note Structured as ghost_targets_data.
+         */
+        std::vector<
+          std::pair<unsigned int, std::pair<unsigned int, unsigned int>>>
+          import_targets_data;
+
+        /**
+         * An array that caches the number of chunks in the import indices per
+         * MPI rank. The length is import_indices_data.size()+1.
+         *
+         * The set of (local) indices that we are importing during compress()
+         * from remote processes, i.e., others' ghosts that belong to the local
+         * range.
+         */
+        std::pair<std::vector<unsigned int>,
+                  std::vector<std::pair<unsigned int, unsigned int>>>
+          import_indices_data;
+
+        /**
+         * Shared-memory ranks from which data is copied from during
+         * export_to_ghosted_array_finish().
+         */
+        std::vector<unsigned int> sm_ghost_ranks;
+
+        /**
+         * Indices from where to copy data from during
+         * export_to_ghosted_array_finish().
+         */
+        std::pair<std::vector<unsigned int>,
+                  std::vector<std::pair<unsigned int, unsigned int>>>
+          sm_export_data;
+
+        /**
+         * Indices where to copy data to during
+         * export_to_ghosted_array_finish().
+         */
+        std::pair<std::vector<unsigned int>,
+                  std::vector<std::pair<unsigned int, unsigned int>>>
+          sm_export_data_this;
+
+        /**
+         * Shared-memory ranks from where to copy data from during
+         * import_from_ghosted_array_finish().
+         */
+        std::vector<unsigned int> sm_import_ranks;
+
+        /**
+         * Indices from where to copy data from during
+         * import_from_ghosted_array_finish().
+         */
+        std::pair<std::vector<unsigned int>,
+                  std::vector<std::pair<unsigned int, unsigned int>>>
+          sm_import_data;
+
+        /**
+         * Indices where to copy data to during
+         * import_from_ghosted_array_finish().
+         */
+        std::pair<std::vector<unsigned int>,
+                  std::vector<std::pair<unsigned int, unsigned int>>>
+          sm_import_data_this;
```

Comment:
These are the new variable names. I still need to refactor how these are set up.

Reply:
I like these names.

Reply:
Puh :) Let me clean up the constructor tomorrow and then this PR is ready from my side!